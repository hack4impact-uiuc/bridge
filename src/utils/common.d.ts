declare module "@hack4impact-uiuc/bridge/common" {
  import { Props, ReactType } from "react";
  import {
    ColorProps,
    SpaceProps,
    LayoutProps,
    TypographyProps as StyledTypographyProps,
    BordersProps as StyledBordersProps,
    BoxShadowProps,
  } from "styled-system";

  export interface BaseProps extends Props<any> {
    as?: ReactType;
    className?: string;
  }
  export interface CommonProps
    extends BaseProps,
      ColorProps,
      SpaceProps,
      LayoutProps {}

  export interface TypographyProps extends BaseProps, StyledTypographyProps {}

  export interface BorderProps extends StyledBordersProps, BoxShadowProps {}
}
