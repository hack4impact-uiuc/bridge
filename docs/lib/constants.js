export const COMPONENTS_URL = '/components';
export const DESIGN_URL = '/design';
