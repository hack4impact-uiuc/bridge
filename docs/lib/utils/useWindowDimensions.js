import { useState, useEffect } from 'react';

function getWindowDimensions() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      const { innerWidth: width, innerHeight: height } = window;
      setWindowDimensions({ width, height });
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}
