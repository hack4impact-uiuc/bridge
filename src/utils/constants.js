import * as styledSystem from 'styled-system';
import systemPropTypes from '@styled-system/prop-types';
import PropTypes from 'prop-types';


const { compose, system } = styledSystem;

// Text Decoration Prop
export const textDecoration = system({
  prop: 'textDecoration',
  cssProperty: 'text-decoration',
});
textDecoration.propTypes = PropTypes.string;


// common style props to be passed into components
export const COMMON = compose(styledSystem.space, styledSystem.layout);
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.layout,
};

// for easier access
export const TYPOGRAPHY = styledSystem.typography;
export const LAYOUT = styledSystem.layout;
export const POSITION = styledSystem.position;
export const FLEX = styledSystem.flexbox;
export const GRID = styledSystem.grid;
export const COLOR = styledSystem.color;
export const BORDER = styledSystem.border;

TYPOGRAPHY.propTypes = systemPropTypes.typography;
LAYOUT.propTypes = systemPropTypes.layout;
POSITION.propTypes = systemPropTypes.position;
FLEX.propTypes = systemPropTypes.flexbox;
GRID.propTypes = systemPropTypes.grid;
COLOR.propTypes = systemPropTypes.color;
BORDER.propTypes = systemPropTypes.border;
