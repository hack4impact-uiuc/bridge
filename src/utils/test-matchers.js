/* eslint-disable react/forbid-foreign-prop-types */
// taken from https://github.com/primer/components/blob/master/src/utils/test-matchers.js
import 'jest-styled-components';

const stringify = (d) => JSON.stringify(d, null, '  ');
/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap'];

// eslint-disable-next-line no-undef
expect.extend({
  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes));
    const expectedPropKeys = Object.keys(propNames.propTypes);
    const missing = expectedPropKeys.filter((key) => !propKeys.has(key))
      .filter((key) => !ALIAS_PROP_TYPES.includes(key));
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`,
    };
  },
});
