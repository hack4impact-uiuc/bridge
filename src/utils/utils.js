import themeGet from '@styled-system/theme-get';
import * as styledSystem from 'styled-system';
import theme from '../theme';

const { get: getKey } = styledSystem;

export const get = (key) => themeGet(key, getKey(theme, key));

export const getOrDefault = (val, def) => (val !== undefined && val !== null ? val : def);

export const lodashGet = (object, path, defaultVal) => {
  const PATH = Array.isArray(path)
    ? path
    : path.split('.').filter((i) => i.length);
  if (!PATH.length) {
    return object === undefined ? defaultVal : object;
  }
  if (object === null || object === undefined || typeof (object[PATH[0]]) === 'undefined') {
    return defaultVal;
  }
  return lodashGet(object[PATH.shift()], PATH, defaultVal);
};
