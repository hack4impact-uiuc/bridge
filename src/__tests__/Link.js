/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';

import { Link } from '..';
import { COMMON, TYPOGRAPHY } from '../utils/constants';
import { renderJSON } from '../utils/testing';

describe('Link', () => {
  it('implements system props', () => {
    expect(Link).toImplementSystemProps(COMMON);
    expect(Link).toImplementSystemProps(TYPOGRAPHY);
  });

  it('renders children', () => {
    const linkText = 'APPLY';
    render(<Link>{linkText}</Link>);
    expect(screen.getByText(linkText)).toBeInTheDocument();
  });

  it('passes href down to link element', () => {
    expect(renderJSON(<Link href="https://uiuc.hack4impact.org" />)).toMatchSnapshot();
  });

  it('renders without any props', () => {
    expect(renderJSON(<Link />)).toMatchSnapshot();
  });

  it('respects hoverColor prop', () => {
    expect(renderJSON(<Link hoverColor="#065535" />)).toMatchSnapshot();
  });

  it('respects color prop', () => {
    expect(renderJSON(<Link color="#065535" />)).toMatchSnapshot();
    const { container } = render(<Link color="#065535" />);
    expect(container.firstChild).toHaveStyle('color: #065535');
  });

  it('defaults to color prop if color & variant are given', () => {
    expect(renderJSON(<Link color="#065535" variant="primary" />)).toMatchSnapshot();
    const { container } = render(<Link color="#065535" />);
    expect(container.firstChild).toHaveStyle('color: #065535');
  });

  it('defaults to hoverColor prop if hoverColor & variant are given while color stays with variant color', () => {
    expect(renderJSON(<Link hoverColor="#065535" variant="white" />)).toMatchSnapshot();
    const { container } = render(<Link hoverColor="#065535" variant="white" />);
    expect(container.firstChild).toHaveStyle('color: #FFFFFF');
  });

  it('respects the "fontStyle" prop', () => {
    const { container } = render(<Link fontStyle="italic" />);
    expect(container.firstChild).toHaveStyle('font-style: italic');
  });
});
