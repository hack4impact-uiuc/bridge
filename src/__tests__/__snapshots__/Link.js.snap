// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link defaults to color prop if color & variant are given 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #065535;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #065535;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

<a
  className="c0"
  color="#065535"
/>
`;

exports[`Link defaults to hoverColor prop if hoverColor & variant are given while color stays with variant color 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #FFFFFF;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #FFFFFF;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #065535;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #065535;
}

<a
  className="c0"
  color="#FFFFFF"
/>
`;

exports[`Link passes href down to link element 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #155DA1;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #155DA1;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

<a
  className="c0"
  color="#155DA1"
  href="https://uiuc.hack4impact.org"
/>
`;

exports[`Link renders without any props 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #155DA1;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #155DA1;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

<a
  className="c0"
  color="#155DA1"
/>
`;

exports[`Link respects color prop 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #065535;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #065535;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

<a
  className="c0"
  color="#065535"
/>
`;

exports[`Link respects hoverColor prop 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #155DA1;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #155DA1;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #065535;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #065535;
}

<a
  className="c0"
  color="#155DA1"
/>
`;

exports[`Link respects the "as" prop with non-string component types 1`] = `
<a
  className="test-link"
>
  Testing...
</a>
`;

exports[`Link should have themed button styling if as="button" 1`] = `
.c0 {
  font-family: inherit;
  font-weight: inherit;
  font-size: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: #155DA1;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-family: Chivo;
  font-weight: 700px;
  font-size: 14px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  text-transform: uppercase;
  display: inline-block;
  padding: 0;
  white-space: nowrap;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background-color: transparent;
  border: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  color: #155DA1;
}

.c0:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

.c0:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
  color: #0E4E8A;
}

<button
  className="c0"
  color="#155DA1"
>
  APPLY
</button>
`;
